zcash_client_backend/src/data_api/chain.rs

```diff
@@ -114,13 +115,71 @@ pub trait BlockSource {
     fn with_blocks<F, WalletErrT>(
         &self,
         from_height: Option<BlockHeight>,
-        limit: Option<u32>,
+        limit: Option<usize>,
         with_row: F,
     ) -> Result<(), error::Error<WalletErrT, Self::Error>>
     where
         F: FnMut(CompactBlock) -> Result<(), error::Error<WalletErrT, Self::Error>>;
 }
 
+/// Verifies that the blocks at the wallet database's view of the chain tip correspond to
+/// blocks available from the block source and form a valid chain that connects to the
+/// `CompactBlock`s in the block_source database.
+///
+/// This function is built on the core assumption that the information provided in the
+/// block source is more likely to be accurate than the previously-scanned information.
+/// This follows from the design (and trust) assumption that the `lightwalletd` server
+/// provides accurate block information as of the time it was requested.
+///
+/// Arguments:
+/// - `block_source` Source of compact blocks
+/// - `data_db` Source of wallet chain tip data
+/// - `validation_depth` the number of blocks prior to the chain tip to validate against
+///   blocks available from the block source.
+/// - `limit` the maximum number of blocks that will be validated. If `None`
+///   is provided, there will be no limit set to the validation and upper bound of the
+///   validation range will be the latest height present in the `block_source`.
```

Comment:
It is unclear how this would actually be used to validate in multiple calls. Consider removing this parameter, because `validation_depth` already bounds the maximum amount of work done.

Also consider making this a new API with a better name rather than calling it `validate_chain`.